### ISSUE TITLE
Annotation for supporting Polymorphism

### ISSUE THREAD

**Issue by ansman** [johncarl81/parceler#139]
I understand that polymorphism isn't supported for performance reasons but it would be nice if you could 
annotate a field so that it checks the class type in runtime. This would also be nice when you have a field with an unmapped generic type.

```java
public class Foo<T> {
    @CheckClass
    public T t;
}

public class Bar {
    @CheckClass
    public SomeObject object;
}

**Comment by johncarl81** [comment]
Hmm, is there a way to handle this with a converter?

**Comment by ansman** [comment]
Sure, I guess it could be done using `Parcels.wrap`

**Comment by cody1024d** [comment]
I'm struggling with a similar issue right now.  The issue with leveraging a custom converter is in the fromParcel method.  Here, you would need to first read the parcelable from the Parcel, in order to pass it into `Parcels.unwrap`, but, in my case, I do not know the actual class in order to pass a class loader.

And yes, unfortunately, the fieldValue has to be an object, as it could be anything from an Enum, to a String, to a class I have created.  They are, however, all parcelable in one way or another through the Parceler framework.

Here is the relevant class:

```
public class SearchFilter {
    protected SysLayoutSearchComparator comparator;
    protected String fieldName;
    @ParcelPropertyConverter(FieldValueConverter.class)
    protected Object fieldValue;

    public SysLayoutSearchComparator getComparator() {
        return comparator;
    }

    public void setComparator(SysLayoutSearchComparator comparator) {
        this.comparator = comparator;
    }

    public String getFieldName() {
        return fieldName;
    }

    public void setFieldName(String fieldName) {
        this.fieldName = fieldName;
    }

    public Object getFieldValue() {
        return fieldValue;
    }

    public void setFieldValue(Object fieldValue) {
        this.fieldValue = fieldValue;
    }

    public static class FieldValueConverter implements TypeRangeParcelConverter {

        @Override
        public void toParcel(Object input, android.os.Parcel parcel) {
            if(input instanceof Parcelable) {
                parcel.writeParcelable((Parcelable)input, 0);
            } else {
                parcel.writeParcelable(Parcels.wrap(input), 0);
            }
        }

        @Override
        public Object fromParcel(android.os.Parcel parcel) {
            return Parcels.unwrap(parcel);
        }
    }
}
```

**Comment by johncarl81** [comment]
Hmm, I wonder if the problem is with your Converter @cody1024d.  Also, you can use a classloader from the current class to do the trick.  Does this work?:

```java
    public static class FieldValueConverter implements ParcelConverter<Object> {

        @Override
        public void toParcel(Object input, android.os.Parcel parcel) {
            if(input instanceof Parcelable) {
                parcel.writeInt(0);
                parcel.writeParcelable((Parcelable)input, 0);
            } else {
                parcel.writeInt(1);
                parcel.writeParcelable(Parcels.wrap(input), 0);
            }
        }

        @Override
        public Object fromParcel(android.os.Parcel parcel) {
            if(parcel.readInt() == 0) {
                return parcel.readParcelable(FieldValueConverter.class.getClassLoader()));
            } else {
                return Parcels.unwrap(parcel.readParcelable(FieldValueConverter.class.getClassLoader())););
            }
        }
    }
```

**Comment by cody1024d** [comment]
Hmm, I can definitely try.  I ended up using a different solution, where I subclassed the above class, based on the type to expect (StringSearchFilter, EnumSearchFilter, etc.), so that I can have a concrete type for the field value.

When I get some time, I will try the above code, as I did not think of writing an int into the parcel to determine if it's a parcelable, or not, and I believe that may have been my issue.  Thank you for the assitance @johncarl81
